cardano-node/src/Cardano/Node/Run.hs

```diff
@@ -326,20 +354,62 @@ handleSimpleNode p trace nodeTracers npm onKernel = do
 
          when mockValidateDB $ traceWith tracer "Performing DB validation"
 
+  stopOnSlotSynced :: NodeProtocolMode -> MaxSlotNo
+  stopOnSlotSynced (RealProtocolMode cli) = shutdownOnSlotSynced cli
+  stopOnSlotSynced _ = NoMaxSlotNo
+
+  -- | Produce a potential pair of FDs:
+  --  - read end, where EOF is a signal of termination
+  --  - write end, which is means of internal termination handling
+  decideShutdownFds :: NodeProtocolMode -> IO ShutdownFDs
+  decideShutdownFds (RealProtocolMode NodeCLI{shutdownIPC = Just fd}) = pure $ ExternalShutdown fd
+  decideShutdownFds (RealProtocolMode NodeCLI{shutdownOnSlotSynced = MaxSlotNo{}}) = mkInternalShutdown
+  decideShutdownFds (MockProtocolMode NodeMockCLI{mockShutdownOnSlotSynced = MaxSlotNo{}}) = mkInternalShutdown
+  decideShutdownFds _ = pure NoShutdownFDs
+
+  mkInternalShutdown :: IO ShutdownFDs
+  mkInternalShutdown = do
+    (r, w) <- IO.createPipeFd
+    pure $ InternalShutdown (Fd r) (Fd w)
+
+data ShutdownFDs
+  = NoShutdownFDs
+  | ExternalShutdown Fd
+  -- ^ We shall watch this externally-supplied FD for termination EOF.
+  | InternalShutdown Fd Fd
+  -- ^ Pair of FDs for intra-process graceful shutdown termination messaging.
+  --   First FD is the read side as above,
+  --   second FD is the shutdown doorbell to be closed.
+
+sfdsReadSide :: ShutdownFDs -> Maybe Fd
+sfdsReadSide = \case
+  ExternalShutdown r -> Just r
+  InternalShutdown r _w -> Just r
+  _ -> Nothing
+
+triggerShutdown :: ShutdownFDs -> Trace IO Text -> Text -> IO ()
+triggerShutdown (InternalShutdown _ (Fd shutFd)) trace reason = do
```

Comment:
What if you just pass it the `Fd`, then you wouldn't have the partiality?

Reply:
Well, then we wouldn't be operating at the level of the abstraction.. and we would introduce ability to express an invalid situation -- we could pass an `Fd` that is not associated with the shutdown doorbell.

Reply:
What if you wrap that `Fd` in a newtype? Or you can make this function a local helper instead of a top-level one.

Reply:
Good point!  I've actually thought of the `newtype` but was a bit afraid of over-engineering -- wasn't sure if this abstraction is worth adding _two_ types..

Reply:
Done.

Reply:
@mrBliss, done -- also newtyped the listener side as well, for symmetry and completeness.  Three types for the shutdown API..